source/common/router/upstream_request.cc

```diff
@@ -416,10 +416,18 @@ void UpstreamRequest::onPoolReady(
     }
   }
 
-  upstream_->encodeHeaders(*parent_.downstreamHeaders(), shouldSendEndStream());
-
+  const Http::Status status =
+      upstream_->encodeHeaders(*parent_.downstreamHeaders(), shouldSendEndStream());
   calling_encode_headers_ = false;
 
+  if (!status.ok()) {
```

Comment:
Can we fail here if we check in FM?  I'd be inclined to remove the asserts here and have router sendLocalReply if status is not Ok (unless it gets hairy to do)

Reply:
If we check in FM, then this won't be triggered (unless something between filter iteration and encoding modifies headers).

>  have router sendLocalReply if status is not Ok (unless it gets hairy to do)

Yes!!! I had a lack of foresight, I can access the decoder callbacks to send local reply through the router. On it!

Reply:
Sorry now that we are checking in the FM, how can this happen? I suppose it can happen if a filter removes headers not in decode headers, but before an async continuation?

Reply:
Yeah, this is my dilemma :/ the FM checks seems necessary and are also sufficient, unless another extension besides filters are able to modify headers filters or encode untrusted headers. so the discussions thus far were

(1) Check per iteration in FM. Would be sufficient right now and robust if filter 1 removes headers and filter 2 dereferences.
(2) Check in encode headers and trust filters use safe accessors like getPathValue. Would be robust to future header modifications.

Reply:
I don't think they are entirely sufficient, as there is nothing to prevent a filter from removing the headers before trying to encode (either directly in something like the router or in some type of async continuation after decodeHeaders() has been checked).

IMO we should definitely do (2). I would be happy to drop (1) for simplicity and performance sake as I'm not sure the per-iteration check actually checks all cases.

Reply:
sounds good, you're right, it probably has minimal benefits and only impacts a filter that was written poorly (with nullptr dereferences rather than safe accessors)! I'll push the change.

Reply:
Done! If the use-case is ever there for per-filter validations or checks, I can answer the call.